app/src/main/java/com/example/dogadoption/repository/DogRepository.kt

```diff
@@ -0,0 +1,32 @@
+package com.example.dogadoption.repository
+
+import android.app.Application
+import com.example.dogadoption.retrofit.DogApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class DogRepository @Inject constructor(
+    private val dogApi: DogApi,
+    private val appContext: Application
+) {
+    suspend fun getDogBreeds(): List<String> {
+        return dogApi.getDogBreeds().message.keys.toList()
+    }
+
+    suspend fun getDogPictures(breed: String): List<String> {
+        return withContext(Dispatchers.IO) {
+            try {
+                val response = dogApi.getDogPictures(breed)
+                if (response.isSuccessful) {
+                    val images = response.body()?.images ?: emptyList()
+                    images.map { it.url }
+                } else {
+                    throw Exception("Failed to fetch dog pictures: ${response.message()}")
+                }
+            } catch (e: Exception) {
+                emptyList()
```

Comment:
Here if the request failed you: `throw error -> you do nothing with it -> you return emptyList` the issue is that you wont know you error occurred. You can log it in the catch block, or..

The repository layer is a perfect place to use the koltin's Result for catching errors. Instead of the try/catch block you can use:

```kotlin
          runCatching {
                val response = dogApi.getDogPictures(breed)
                if (response.isFailure) error("Failed to fetch dog pictures: ${response.message()}")

                val body = response.body() ?: error("Missing body")
                // if `images` is also nullable then you can do another `val images = body.images ?: error("Missing images")
                body.images.map { it.url } 
          }
```

then in the ViewModel you will be able to use it like this:
```kotlin
getDogPictures(breed).fold(
    success = {
        // Update your dogs list with new dogs
    },
    failure = { t ->
        // All those `error()` got here and you can read them `t.message`
        // Do something with the error. Log it or send it to Toast or Snackbar
    },
)
```